File: design-patterns/ts/prototype/transport/car.ts
// car.ts

import Prototype from "./prototype";

class Car implements Prototype {
    make: number;
    model: string;
    color: string;

    constructor(make: number, model: string, color: string) {
        this.make = make;
        this.model = model;
        this.color = color;
    }

    clone(): Prototype {
        return new Car(this.make, this.model, this.color);
    }

    toString(): string {
        return "Make: " + this.make + " | Model: " + this.model + " | Color: " + this.color;
    }
}

export default Car;

File: design-patterns/go/prototype/transport/car.go
// car.go

package main

import "fmt"

type Car struct {
	Make  int
	Model string
	Color string
}

func NewCar() (car *Car) {
	car = &Car{}
	return
}

func (car *Car) Clone() Prototype {
	carClone := NewCar()
	carClone.Make = car.Make
	carClone.Model = car.Model
	carClone.Color = car.Color

	return carClone
}

func (car *Car) ToString() string {
	return fmt.Sprintf("Make: %v  | Model: %v | Color: %v", car.Make, car.Model, car.Color)
}
